### PR TITLE
CS-36: Create rspec matcher for dependency_container's import

### PR DESCRIPTION
**Pull request by Olha-30**
## What was done as a part of this PR?

- Created the import matcher

## Why it was done?

- To check the imported methods from the dependency container

## How to test?
- `task rspec -- spec/lib/convenient_service/rspec/matchers/custom/import_spec.rb`
- [Development: Local Development#tests](https://github.com/marian13/convenient_service/wiki/Development:-Local-Development#tests).

### Usage example
- Let's imagine you have a container like this one

```ruby
module Container
  include ConvenientService::DependencyContainer::Export

  export :"foo::bar::baz", scope: :class do 
    "foo::bar::baz with scope: :class"
  end

  export :qux do
    ":qux with scope: :instance"
  end

  export :"quux.xyzzy", scope: :class do
    ":quux.xyzzy with scope: :class"
  end
end
```
- And the class that imports methods from the `Container`
```ruby
class Service
  include ConvenientService::DependencyContainer::Import

  import :"foo::bar::baz", scope: :class, from: Container
  
  import :qux, from: Container

  import :"quux.xyzzy", as: :xyzzy, scope: :class, from: Container, prepend: true
end
```
- How the specs will look like
```ruby
RSpec.describe Service do
  include ConvenientService::RSpec::Matchers::Import

  subject { described_class }

  # you can skip the `scope` keyword for `instance` as it's the default value
  # you can skip the `prepend` keyword for `false` as it's the default value

  it { is_expected.to import(:"foo::bar::baz", scope: :class, from: Container) }

  it { is_expected.to import(:qux, from: Container) }

  it { is_expected.to import(:"quux.xyzzy", as:  :xyzzy, scope: :class, from: Container, prepend: true) }
end
```

## Checklist:

- [x] I have performed a self-review of my code.
- [x] I have added/adjusted tests that prove my fix is effective or that my feature works.
- [x] CI is green.

### CODE REVIEW

File: lib/convenient_service/rspec/matchers/custom/import.rb [hunk 1]
```diff
@@ -0,0 +1,148 @@
+# frozen_string_literal: true
+
+module ConvenientService
+  module RSpec
+    module Matchers
+      module Custom
+        class Import
+          ##
+          # @param method_name [Symbol, String]
```
Comment (marian13) [inline lib/convenient_service/rspec/matchers/custom/import.rb:9]: **Expressiveness**

`full_name`.

File: lib/convenient_service/rspec/matchers/custom/import.rb [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+          # @param scope [Symbol]
+          # @param from [Class, Module]
```
Comment (marian13) [inline lib/convenient_service/rspec/matchers/custom/import.rb:11]: Looks like `DependencyContainers` have incomplete behavior.

Could you please, implement a logic that forbids to use of `export` in classes?

Thanks.

Comment (Olha-30) [reply lib/convenient_service/rspec/matchers/custom/import.rb:11]: @marian13 Check the following PR #61 &#127881; &#127881; &#127881;

File: lib/convenient_service/rspec/matchers/custom/import.rb [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+          ##
+          # @param slug [Symbol, String]
+          # @param scope [Symbol, Constants::DEFAULT_SCOPE]
+          # @param from [Module]
+          # @param prepend [Boolean, Constants::DEFAULT_PREPEND]
+          # @param as [Symbol, String]
+          # @return [void]
+          #
+          def initialize(slug, from:, as: "", scope: constants.DEFAULT_SCOPE, prepend: constants.DEFAULT_PREPEND)
```
Comment (marian13) [inline lib/convenient_service/rspec/matchers/custom/import.rb:25]: `as: ""` -> `as: constants.DEFAULT_ALIAS`

File: lib/convenient_service/rspec/matchers/custom/import.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+          def failure_message_when_negated
+            "expected `#{klass}` NOT to import `#{slug}`#{string_to_print_depending_on_slug} with scope: `#{scope}` from: `#{from}` with  prepend: `#{prepend}`"
+          end
+
+          private
+
+          ##
+          # @return [String]
+          #
+          def string_to_print_depending_on_slug
```
Comment (marian13) [inline lib/convenient_service/rspec/matchers/custom/import.rb:79]: `def string_to_print_depending_on_slug` -> `def printable_alias_slug`

File: lib/convenient_service/rspec/matchers/custom/import.rb [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+          ##
+          # @param klass [Class, Module]
+          # @return [Boolean]
+          #
+          def matches?(klass)
+            commands.AssertValidContainer.call(container: from)
+
+            commands.AssertValidScope.call(scope: scope)
+
```
Comment (marian13) [inline lib/convenient_service/rspec/matchers/custom/import.rb:41]: commands.AssertValidSlug.call(slug: slug)

File: lib/convenient_service/rspec/matchers/custom/import.rb [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            "expected `#{klass}` NOT to import `#{slug}`#{string_to_print_depending_on_slug} with scope: `#{scope}` from: `#{from}` with  prepend: `#{prepend}`"
+          end
+
+          private
+
+          ##
+          # @return [String]
+          #
+          def string_to_print_depending_on_slug
+            alias_slug.to_s.empty? ? "" : " as: `#{alias_slug}`"
```
Comment (marian13) [inline lib/convenient_service/rspec/matchers/custom/import.rb:80]: Please, remove `.to_s`

File: lib/convenient_service/support/dependency_container/commands/import_method.rb [hunk 7]
```diff
@@ -57,24 +57,7 @@ def call
...
-
-          ##
-          # @return [String]
-          #
-          def scoped_prefix
-            case scope
-            when Constants::INSTANCE_SCOPE then "Instance"
-            when Constants::CLASS_SCOPE then "Class"
-            end
+            @imported_scoped_methods ||= Commands::FetchImportedScopedMethods.call(importing_module: importing_module, scope: scope, prepend: prepend) { Commands::CreateMethodsModule.call }
```
Comment (marian13) [inline lib/convenient_service/support/dependency_container/commands/import_method.rb:60]:
```ruby
@imported_scoped_methods ||= Commands::FetchImportedScopedMethods.call(importing_module: importing_module, scope: scope, prepend: prepend)
```

File: lib/convenient_service/support/dependency_container/entities/method.rb [hunk 8]
```diff
@@ -38,7 +38,7 @@ class Method
           # @param alias_slug [String, Symbol]
           # @return [void]
           #
-          def initialize(slug:, scope:, body:, alias_slug: "")
+          def initialize(slug:, scope:, body: nil, alias_slug: "")
```
Comment (marian13) [inline lib/convenient_service/support/dependency_container/entities/method.rb:41]: `body: proc {}`

File: lib/convenient_service/support/dependency_container/commands/fetch_imported_scoped_methods.rb [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+
+          ##
+          # @return [Module]
+          #
+          def call
+            Utils::Module.fetch_own_const(importing_module, :"Imported#{imported_prefix}#{scoped_prefix}Methods", &block)
+          end
+
+          private
+
```
Comment (marian13) [inline lib/convenient_service/support/dependency_container/commands/fetch_imported_scoped_methods.rb:54]: BuildNamespaceName
# symbol :"Imported#{imported_prefix}#{scoped_prefix}Methods"

FetchNamespace
# const

GetNamespace
# const





def namespace_name
   @namespace_name ||= Commans::BuildNamespaceName.call()
end

File: lib/convenient_service/support/dependency_container/commands/get_namespace.rb [hunk 10]
```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+module ConvenientService
+  module Support
+    module DependencyContainer
+      module Commands
+        class GetNamespace < Support::Command
```
Comment (marian13) [inline lib/convenient_service/support/dependency_container/commands/get_namespace.rb:7]: `ExtractNamespace`

File: lib/convenient_service/support/dependency_container/entities/method.rb [hunk 11]
```diff
@@ -34,11 +34,11 @@ class Method
           ##
           # @param slug [String, Symbol]
           # @param scope [:instance, :class]
-          # @param body [Proc]
-          # @param alias_slug [String, Symbol]
+          # @param body [Proc, ConvenientService::Support::DependencyContainer::Constants::DEFAULT_BODY]
```
Comment (marian13) [inline lib/convenient_service/support/dependency_container/entities/method.rb:37]:
```ruby
# @param body [Proc]
```

File: lib/convenient_service/support/dependency_container/entities/method.rb [hunk 12]
```diff
@@ -128,14 +149,30 @@ def to_kwargs
...
+            methods_depending_on_namespace_class(namespace).find { |method| method == name }
+          end
+
+          ##
+          # @return [Array]
+          #
+          def methods_depending_on_namespace_class(namespace)
+            return namespace.instance_methods(false) if namespace.instance_of?(::Module)
+
+            namespace.singleton_methods(false)
```
Comment (marian13) [inline lib/convenient_service/support/dependency_container/entities/method.rb:175]:
```ruby
##
# @param include_super [Boolean]
#
def instance_methods(include_super = true)
  singleton_methods(include_super)
end
```

File: lib/convenient_service/support/dependency_container/entities/method.rb [hunk 13]
```diff
@@ -95,6 +106,16 @@ def define_in_module!(mod)
...
 
+          def defined_in_module?(mod)
+            method = processed_method_slug_parts.reduce(mod) do |namespace, name|
+              next namespace unless namespace
+
+              namespace.namespaces.find_by(name: name) || find_method(namespace, name)
+            end
+
+            method == name
+          end
```
Comment (marian13) [inline lib/convenient_service/support/dependency_container/entities/method.rb:117]:
```ruby
##
# @internal
#   NOTE: Consider we have the following slug - "foo.bar.baz.qux.quux" then the innermost namespace is `qux`.
#
def find_innermost_namespace(mod)
  expected_namespaces.reduce(mod) do |namespace, sub_namespace|
    break unless namespace

    namespace.namespaces.find_by(name: sub_namespace.name)
  end
end

##
# @internal
#   NOTE: Consider we have the following slug - "foo.bar.baz.qux.quux" then method is `quux`.
#
def defined_in_module?(mod)
  innermost_namespace = find_innermost_namespace(mod)

  return false unless innermost_namespace

  method = innermost_namespace.instance_methods(false).find { |method_name| method_name == name }

  Utils.to_bool(method)
end
```

File: spec/lib/convenient_service/rspec/matchers/custom/import_spec.rb [hunk 14]
```diff
@@ -0,0 +1,286 @@
...
+        end
+
+        it "returns true" do
+          expect(matcher_result).to be_truthy
+        end
+      end
+
+      context "when `namespace` is NOT present" do
+        it "returns false" do
+          expect(matcher_result).to be_falsey
```
Comment (marian13) [inline spec/lib/convenient_service/rspec/matchers/custom/import_spec.rb:152]:
```ruby
expect(matcher_result).to eq(false)
```

File: spec/lib/convenient_service/rspec/matchers/custom/import_spec.rb [hunk 15]
```diff
@@ -0,0 +1,286 @@
...
+          .with_arguments(importing_module: klass, scope: scope, prepend: prepend)
+      end
+
+      context "when `namespace` is present" do
+        before do
+          klass.import(imported_slug, **kwargs)
+        end
+
+        it "returns true" do
+          expect(matcher_result).to be_truthy
```
Comment (marian13) [inline spec/lib/convenient_service/rspec/matchers/custom/import_spec.rb:146]:
```ruby
expect(matcher_result).to eq(true)
```

**Pull request closed without merging**